### PR TITLE
Android: Implement custom driver loading for ARM64 Android devices

### PR DESCRIPTION
**Pull request by dima-xd**
Add possibility to install custom Vulkan driver for ARM64 Android devices using [libadrenotools](https://github.com/bylaws/libadrenotools).
Some drivers can be found [here](https://github.com/K11MCH1/AdrenoToolsDrivers/releases/).
It is also possible to use this [feature](https://github.com/bylaws/libadrenotools/blob/master/include/adrenotools/driver.h#L60), but I'm not sure if it's actually needed.

~~**Draft for now, until I figure out why some drivers are crashing [here](https://github.com/hrydgard/ppsspp/blob/master/Common/GPU/Vulkan/VulkanQueueRunner.cpp#L420). Or until I figure out  how to make some kind of fallback to use default if it crashes on booting the app.**~~

### CODE REVIEW

**Comment by hrydgard** [comment]
Crash in debug layer? You might be able to work around it in the Vulkan initialization by disabling the `VK_EXT_debug_utils` extension when you load these drivers, then we won't try to call those functions.

File: Core/Config.h
Comment (hrydgard) [inline Core/Config.h:150]: indentation

**Comment by hrydgard** [comment]
We can play with the turbo setting and stuff in followup PRs, don't need to get everything in at once.

**Comment by dima-xd** [comment]
> Crash in debug layer? You might be able to work around it in the Vulkan initialization by disabling the `VK_EXT_debug_utils` extension when you load these drivers, then we won't try to call those functions.

Added the check if `customDriver` is empty. Works as it should now. Thanks for the advice

**Comment by hrydgard** [comment]
Are you seeing any meaningful improvements with the driver, by the way?

**Comment by dima-xd** [comment]
> Are you seeing any meaningful improvements with the driver, by the way?



> Are you seeing any meaningful improvements with the driver, by the way?

I didn't test too much. For some reason Qualcomm drivers like 615.37 work very wierdly (hard to explain, better to see it yourself). But with turnip drivers I get like +5-20% performance, so I don't even need to apply speedhacks. But it' only in games I tested.

**Comment by hrydgard** [comment]
That sounds promising, at least. Though, even my old Poco F1 with Adreno 630 runs all games at full speed with the proprietary drivers already...

**Comment by dima-xd** [comment]
> That sounds promising, at least. Though, even my old Poco F1 with Adreno 630 runs all games at full speed with the proprietary drivers already... but it's good to have this to experiment with.

Hmm... Almost the same for me with my Poco F3. But anyway, is there any benchmark homebrew/game to understand better performance boost (want to test with turbo mode too)?

**Comment by hrydgard** [comment]
I think the most interesting might be to try with really low-end Adreno devices that still allow driver replacements. Unfortunately I don't have any of those.

Maybe crank the resolution to see if the fragment shaders get more efficient?

Some relatively heavy games to run are the God of War games, Outrun, Wipeout Pure/Pulse, Tekken 6, GTA LCS / VCS...

BTW, does this driver has a way to set a stable power state (avoiding all the up and down clocking, setting a fixed slow clock for example)?

**Comment by dima-xd** [comment]
> BTW, does this driver has a way to set a stable power state (avoiding all the up and down clocking, setting a fixed slow clock for example)?

not sure about setting a fixed slow clock, I can ask the dev of this lib. But if's possible to set stable high clocks with turbo mode, then probably it also possible to do it reversed

**Comment by hrydgard** [comment]
Just asking because it may be nice to have when optimizing shaders, to avoid false measurements due to the clock changing.

Our shaders are quite simple though generally.

**Comment by hrydgard** [comment]
How many times are you gonna force-push? :) Since this is your first PR, I have to approve each CI run manually :)

**Comment by dima-xd** [comment]
> How many times are you gonna force-push? :) Since this is your first PR, I have to approve each CI run manually :)

oops, sorry :) I guess I'll wait for CI and try to force-push last time

**Comment by hrydgard** [comment]
hm, seem to be some include path issue?

**Comment by dima-xd** [comment]
> hm, seem to be some include path issue?

yep, not sure why. Locally it works fine

File: Common/GPU/Vulkan/VulkanContext.cpp
Comment (hrydgard) [inline Common/GPU/Vulkan/VulkanContext.cpp:136]: This expression is better written with parenthesis around the & operator subexpression. I'm always suspicious of operator precedence around binary operators vs logical ones..

**Review by hrydgard — CHANGES_REQUESTED** [review]
Oh right, I know what it is. On CI here we build without gradle, instead we use a legacy Android.mk for, well, legacy reasons.

So, you'll need to add any new include paths and new cpp files to that as well.

I'd suggest adding include paths here: 

https://github.com/hrydgard/ppsspp/blob/master/android/jni/Locals.mk#L24

**Comment by dima-xd** [comment]
> Oh right, I know what it is. On CI here we build without gradle, instead we use a legacy Android.mk for, well, legacy reasons.
> 
> So, you'll need to add any new include paths and new cpp files to that as well.
> 
> I'd suggest adding include paths here:
> 
> https://github.com/hrydgard/ppsspp/blob/master/android/jni/Locals.mk#L24

I see. It's already late, so I'll do that tomorrow.
Also with gradle I'll need to add `jniLibs.useLegacyPackaging = true` to 'packagingOptions', because iirc drivers won't work on release builds. So I hope there won't be any problems with make since it doesn't use gradle

**Comment by hrydgard** [comment]
Won't work on release builds? That sounds weird.

We definitely still need to be able to make release builds, for Google Play. So we need to gate the driver loading feature from them?

**Comment by dima-xd** [comment]
> Won't work on release builds? That sounds weird.
> 
> We definitely still need to be able to make release builds, for Google Play. So we need to gate the driver loading feature from them?

Ehh, maybe I didn't explain well. What I want to say is that custom driver loading won't work on release builds without `jniLibs.useLegacyPackaging = true`. So basically I need to add this option to gradle and loading will work on release builds too. But I am not sure how this works with Make since I never used it.
Probably Make doesn't even need this option to be specified, but I am not 100% sure

**Comment by hrydgard** [comment]
Right, you can ignore it in Make since those builds are not actually used.

**Comment by dima-xd** [comment]
Let's hope it will compile now

**Comment by dima-xd** [comment]
> Maybe crank the resolution to see if the fragment shaders get more efficient?
> 

Didn't find any difference on my device

> BTW, does this driver has a way to set a stable power state (avoiding all the up and down clocking, setting a fixed slow clock for example)?

The dev of this lib said that's impossible, unfortunately

**Comment by Cristobal15** [comment]
Dima XD. MotorStorm: Artic Edge and Porsuite Force Extreme Justice are among the games that Android devices struggle with the most. The SD865 has a hard time getting it up to 3X. They would be good test games.

**Comment by dima-xd** [comment]
> Dima XD. MotorStorm: Artic Edge and Porsuite Force Extreme Justice are among the games that Android devices struggle with the most. The SD865 has a hard time getting it up to 3X. They would be good test games.

MotorStorm: Artic Edge with x4 resolution with Mesa driver seems like 1-3 fps is faster VS default driver.
But it's hard to measure. It's better to test with lower end devices

**Comment by hrydgard** [comment]
`../../Common/GPU/Vulkan/VulkanLoader.cpp:316: undefined reference to `adrenotools_open_libvulkan'`

Think we need to add any cpp files from the adrenotools library as well to the Android.mk. 

(Yes, I know it's a pain. we should probably switch CI to gradle).

**Comment by dima-xd** [comment]
> `../../Common/GPU/Vulkan/VulkanLoader.cpp:316: undefined reference to `adrenotools_open_libvulkan'`
> 
> Think we need to add any cpp files from the adrenotools library as well to the Android.mk (if any), and add `adrenotools_open_libvulkan` as a library reference.
> 
> (Yes, I know it's a pain. we should probably switch CI to gradle).

You mean something like this? Unfortunately I can't even test it locally, and I don't want to trigger CI a lot

```diff
index 70fda2601..d0a90937b 100644
--- a/android/jni/Android.mk
+++ b/android/jni/Android.mk
@@ -140,6 +140,13 @@ RCHEEVOS_FILES := \
   ${SRC}/ext/rcheevos/src/rhash/hash.c \
   ${SRC}/ext/rcheevos/src/rhash/md5.c

+ADRENOTOOLS_FILES := \
diff --git a/android/jni/Android.mk b/android/jni/Android.mk
index 70fda2601..d0a90937b 100644
--- a/android/jni/Android.mk
+++ b/android/jni/Android.mk
@@ -140,6 +140,13 @@ RCHEEVOS_FILES := \
   ${SRC}/ext/rcheevos/src/rhash/hash.c \
   ${SRC}/ext/rcheevos/src/rhash/md5.c

+ADRENOTOOLS_FILES := \
+  ${SRC}/ext/libadrenotools/src/driver.cpp \
+  ${SRC}/ext/libadrenotools/src/hook/hook_impl.cpp \
+  ${SRC}/ext/libadrenotools/src/hook/file_redirect_hook.c \
+  ${SRC}/ext/libadrenotools/src/hook/gsl_alloc_hook.c \
+  ${SRC}/ext/libadrenotools/src/hook/main_hook.c
+
 VR_FILES := \
   $(SRC)/Common/VR/OpenXRLoader.cpp \
   $(SRC)/Common/VR/PPSSPPVR.cpp \
@@ -212,6 +219,7 @@ EXEC_AND_LIB_FILES := \
   $(SPIRV_CROSS_FILES) \
   $(RCHEEVOS_FILES) \
   $(NAETT_FILES) \
+  $(ADRENOTOOLS_FILES) \
   $(EXT_FILES) \
   $(NATIVE_FILES) \
   $(SRC)/Common/Buffer.cpp \
(END)
   $(SRC)/Common/VR/OpenXRLoader.cpp \
   $(SRC)/Common/VR/PPSSPPVR.cpp \
@@ -212,6 +219,7 @@ EXEC_AND_LIB_FILES := \
   $(SPIRV_CROSS_FILES) \
   $(RCHEEVOS_FILES) \
   $(NAETT_FILES) \
+  $(ADRENOTOOLS_FILES) \
   $(EXT_FILES) \
   $(NATIVE_FILES) \
   $(SRC)/Common/Buffer.cpp \
   $(SRC)/Common/VR/PPSSPPVR.cpp \
@@ -212,6 +219,7 @@ EXEC_AND_LIB_FILES := \
   $(SPIRV_CROSS_FILES) \
   $(RCHEEVOS_FILES) \
   $(NAETT_FILES) \
+  $(ADRENOTOOLS_FILES) \
   $(EXT_FILES) \
   $(NATIVE_FILES) \
   $(SRC)/Common/Buffer.cpp \
   $(SRC)/Common/VR/PPSSPPVR.cpp \
@@ -212,6 +219,7 @@ EXEC_AND_LIB_FILES := \
   $(SPIRV_CROSS_FILES) \
   $(RCHEEVOS_FILES) \
   $(NAETT_FILES) \
+  $(ADRENOTOOLS_FILES) \
   $(EXT_FILES) \
   $(NATIVE_FILES) \
   $(SRC)/Common/Buffer.cpp \
   $(SRC)/Common/VR/PPSSPPVR.cpp \
@@ -212,6 +219,7 @@ EXEC_AND_LIB_FILES := \
   $(SPIRV_CROSS_FILES) \
   $(RCHEEVOS_FILES) \
   $(NAETT_FILES) \
+  $(ADRENOTOOLS_FILES) \
   $(EXT_FILES) \
   $(NATIVE_FILES) \
   $(SRC)/Common/Buffer.cpp \
 ^X
   $(SRC)/Common/VR/PPSSPPVR.cpp \
@@ -212,6 +219,7 @@ EXEC_AND_LIB_FILES := \
   $(SPIRV_CROSS_FILES) \
   $(RCHEEVOS_FILES) \
   $(NAETT_FILES) \
+  $(ADRENOTOOLS_FILES) \
   $(EXT_FILES) \
   $(NATIVE_FILES) \
   $(SRC)/Common/Buffer.cpp \
   ````

**Comment by hrydgard** [comment]
I can run the old build locally tomorrow and figure it out.

**Comment by dima-xd** [comment]
> I can run the old build locally tomorrow and figure it out.

I just set up Make in WSL. Will fix building soon

**Comment by dima-xd** [comment]
> `../../Common/GPU/Vulkan/VulkanLoader.cpp:316: undefined reference to `adrenotools_open_libvulkan'`
> 
> Think we need to add any cpp files from the adrenotools library as well to the Android.mk (if any), and add `adrenotools_open_libvulkan` as a library reference.
> 
> (Yes, I know it's a pain. we should probably switch CI to gradle).

This should be it.
I can do migration to gradle one day myself, if you don't mind

**Comment by hrydgard** [comment]
Well, we're gonna wanna talk to Unknown and see if he's still using that path. I don't use it myself anymore

**Comment by sum2012** [comment]
@dima-xd please add translate string for "Current GPU Driver" and "Install Custom Driver..."
in \assets\lang\*.ini

**Comment by hrydgard** [comment]
Just noting that an easy way to add the translation strings is to use langtool, though it requires a rust install.

From Tools/langtool:

cargo run -- add-new-key Section Key

**Comment by Narugakuruga** [comment]
Qualcomm is well known for droppping support for old device. Even a 1 year old phone could be neglected and not receiving any driver update.

So it could be meaningful when PPSSPP need new vulkan extensions for extra features.

And better performance could mean lower power consumption, will benefit the overall user experience.

**Comment by hrydgard** [comment]
Some issues, which I'm happy to help fix after merging:

* Choosing a file that's not a zip file (like, the unzipped .so file) gives no feedback that you picked the wrong file
* It's a bit too prominent, and it's not really clear to a random user what it's asking for.
* Our driver bug detection probably should special case Turnip and turn off our mitigations (hopefully it doesn't have the same bugs)...

So, I'm going to temporarily move this to `Tools/Developer Tools` for now, and later maybe migrate it back to Graphics settings.

But, super cool to have, and works great!

**Review by hrydgard — APPROVED** [review]

**Comment by hrydgard** [comment]
@dima-xd FYI, followup in #18548

**Comment by unknownbrackets** [comment]
> Well, we're gonna wanna talk to Unknown and see if he's still using that path. I don't use it myself anymore

Gotta say, I really can't work my head around how you can stand making a button every time you want to try any little piece of code in PPSSPP.  The idea of not being able to run ppssppunittest and ppssppheadless on Android phones is frustrating to me.  Burns so much more time.

-[Unknown]

**Comment by hrydgard** [comment]
I run them on my Mac instead to get ARM64 coverage. But yeah. I probably would benefit from using that path more heh.

**Comment by kalebepalacio** [comment]
Is there really any advantage to using Turnip on ppsspp? I honestly don't see it, but I would like an official answer (if possible).

**Comment by hrydgard** [comment]
Some people have indeed reported substantial speedups, though it seems to depend on the device and the particular driver it has by default. I'm personally not really convinced this is worth it but it was heavily requested, so ...

**Comment by kalebepalacio** [comment]
I understood. Thanks for the feedback.

**Pull request merged**
